### PR TITLE
Feature UNO-341 Implement remote sourced lists

### PR DESCRIPTION
**Pull request by SergiiTao**
Ticket: [UNO-341](https://oat-sa.atlassian.net/browse/UNO-341)

depends on oat-sa/tao-core#2598

Goal: to add remote sourced lists. so it will save time for an administrator and have a single data source.

Remote sourced lists a fully compatible with default ones.

Screenshot from 2020-07-16 15-33-40

### CODE REVIEW

File: controller/Lists.php [hunk 1]
```diff
@@ -83,15 +88,64 @@ public function index()
...
+
+        if ($remoteListForm->isSubmited()) {
+            if ($remoteListForm->isValid()) {
+                $values = $remoteListForm->getValues();
+
+                $newList = $remoteSourcedListService->createList(
+                    $values[tao_actions_form_RemoteList::FIELD_NAME],
+                    $values[tao_actions_form_RemoteList::FIELD_SOURCE_URL],
+                    $values[tao_actions_form_RemoteList::FIELD_ITEM_LABEL_PATH],
+                    $values[tao_actions_form_RemoteList::FIELD_ITEM_URI_PATH]
```
Comment (shpran) [inline controller/Lists.php:118]: Are you sure `$values` array contains all of these items?

File: model/lists/ListService.php [hunk 2]
```diff
@@ -32,13 +40,81 @@ class ListService extends \tao_models_classes_ListService
...
         return $listClass->isSubClassOf($this->getClass(TaoOntology::CLASS_URI_LIST))
-        && $listClass->getUri() !== \tao_models_classes_LanguageService::CLASS_URI_LANGUAGES;
+            && $listClass->getUri() !== \tao_models_classes_LanguageService::CLASS_URI_LANGUAGES;
+    }
+
+    public function getListElement(RdfClass $listClass, $uri)
+    {
+        $request = new ValueCollectionSearchRequest();
+        $request->setValueCollectionUri($listClass->getUri())//->setObject()
+        ;
```
Comment (shpran) [inline model/lists/ListService.php:59]:
```suggestion
        $request->setValueCollectionUri($listClass->getUri());//->setObject()
```

File: model/lists/ListService.php [hunk 3]
```diff
@@ -32,13 +40,81 @@ class ListService extends \tao_models_classes_ListService
...
+
+        $result = $this->getValueService()->findAll(
+            new ValueCollectionSearchInput($request)
+        );
+
+        return iterator_to_array($result->getIterator())[0];
```
Comment (shpran) [inline model/lists/ListService.php:65]: Are you sure a zero element exists?

File: model/lists/ListService.php [hunk 4]
```diff
@@ -32,13 +40,81 @@ class ListService extends \tao_models_classes_ListService
...
+     * @param RdfClass $listClass
+     *
+     * @return bool
+     * @throws core_kernel_persistence_Exception
+     */
+    public function isRemote(RdfClass $listClass): bool
+    {
+        $type = $listClass->getOnePropertyValue($listClass->getProperty('http://www.tao.lu/Ontologies/TAO.rdf#ListType'));
+
+        return $type && ($type->getUri() === 'http://www.tao.lu/Ontologies/TAO.rdf#ListRemote');
```
Comment (shpran) [inline model/lists/ListService.php:118]: Add/use constants for these values:
* `'http://www.tao.lu/Ontologies/TAO.rdf#ListType'`
* `'http://www.tao.lu/Ontologies/TAO.rdf#ListRemote'`

File: controller/Lists.php (hunk 1 above)
Comment (SergiiTao) [reply controller/Lists.php:128]: `$values`  doesn't depend on request parameters, so I'd say that I am sure

File: controller/Lists.php [hunk 5]
```diff
@@ -149,9 +319,19 @@ public function getListElements()
...
+                $limit    = $isRemote
+                    ? self::REMOTE_LIST_PREVIEW_LIMIT
+                    : 0;
+
+                foreach ($this->getListService()->getListELements($list, true, $limit) as $listElement) {
                     $data[tao_helpers_Uri::encode($listElement->getUri())] = $listElement->getLabel();
                 }
+
+                if ($isRemote && count($data) === self::REMOTE_LIST_PREVIEW_LIMIT) {
+                    $data[''] = '...';
```
Comment (Alroniks) [inline controller/Lists.php:333]: Is the key missed here?

Comment (wazelin) [reply controller/Lists.php:333]: No, it doesn't have any meaning in the context, but all the other keys are string values.

Comment (Alroniks) [reply controller/Lists.php:333]: Yes, about other keys it is clear, but what the reason to add value under empty key?

Comment (wazelin) [reply controller/Lists.php:333]: That's just a quick hack to highlight the fact the list, being shown, is incomplete, and there's more to it.
image

File: controller/Lists.php [hunk 6]
```diff
@@ -130,12 +130,17 @@ public function remote(
...
+                }
+                catch (ValueConflictException $exception) {
+                    return $this->returnError(
+                        $exception->getMessage() . __(' Probably given list was already imported.')
+                    );
+                }
+                catch (RuntimeException $exception) {
                     throw $exception;
+                } finally {
+                    $this->removeList(tao_helpers_Uri::encode($newList->getUri()));
```
Comment (wazelin) [inline controller/Lists.php:142]: Wait, that will be executed always, even without any exception!

Comment (wazelin) [reply controller/Lists.php]: Fixed.

**Pull request merged**